### PR TITLE
Configurable hash algorithm

### PR DESCRIPTION
**Pull request by rprevot**
Dynomite supports several hash algorithms, but Dyno uses implicitly the Murmur hash function. 
This pull request provides configurable HashPartitioner through ConnectionPoolConfiguration.

### CODE REVIEW

**Comment by ipapapa** [comment]
Can you point me where Dynomite provides more than murmur1 hash algorithm? Currently, Dyno provides multiple hash algorithms support including murmur1, murmur2, murmur3, Binary Search etc. For more information: https://github.com/Netflix/dyno/tree/master/dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/hash How is this PR going to improve that?

**Comment by rprevot** [comment]
I refer to the "hash" option in Dynomite [configuration](http://www.dynomitedb.com/docs/dynomite/v0.5.6/dynomite-yaml/). That option is used during [ pool initialization](https://github.com/Netflix/dynomite/blob/4ba783990a620551ae0c119008b62c6b2fad36c2/src/dyn_server.c#L431). The hash functions are defined in [hashkit](https://github.com/Netflix/dynomite/blob/4ba783990a620551ae0c119008b62c6b2fad36c2/src/hashkit/dyn_hashkit.h#L33) subtree.

The sample logs below illustrates the Dynomite behaviour with hash=fnv1a_32 :
```
[2017-09-28 09:53:08.215] conf_push_scalar:541 push 'fnv1a_32'
[2017-09-28 09:53:08.215] conf_handler:1291 conf handler on hash: fnv1a_32
....
[2017-09-28 09:53:08.215] conf_dump:357 dyn_o_mite
[2017-09-28 09:53:08.215] conf_dump:359   listen: 127.0.0.1:8111
[2017-09-28 09:53:08.215] conf_dump:360   timeout: 5000
[2017-09-28 09:53:08.215] conf_dump:361   backlog: 512
[2017-09-28 09:53:08.215] conf_dump:362   hash: 8
```

Dyno provides multiple hash algorithms, but Murmur1 usage is hardcoded in [TokenAwareSelection ](https://github.com/Netflix/dyno/blob/893b2fdb061ef6d6673f03585320c412a3209df5/dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/TokenAwareSelection.java#L51) strategy.

The goal of this PR is to efficiently load balance requests when Dynomite server doesn't use default hashing function. The hash function is configurable through ConnectionPoolConfiguration :

```
ConnectionPoolConfigurationImpl config 
	= new ConnectionPoolConfigurationImpl(cluster.getName())
		.withHashPartitioner(new Fnv1aHashPartitioner())
		.withTokenSupplier(new SimpleTokenMapSupplier(hosts))
		.withErrorMonitorFactory(new SimpleErrorMonitorFactory(errorThreshold));
```

File: dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/HostSelectionWithFallback.java [hunk 1]
```diff
@@ -429,17 +430,23 @@ public void removeHost(Host host, HostConnectionPool<CL> hostPool) {
...
 		@Override
 		public HostSelectionStrategy<CL> vendPoolSelectionStrategy() {
 			
 			switch (lbStrategy) {
 			case RoundRobin:
 				return new RoundRobinSelection<CL>();
-			case TokenAware:
-				return new TokenAwareSelection<CL>();
+			case TokenAware:     
+				return hashPartitioner != null
```
Comment (jcacciatore) [inline dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/HostSelectionWithFallback.java]: Use a WARN level logging statement that a custom hash partitioner is being used

**Comment by ipapapa** [comment]
I am sorry for my delay on this. #186 is a fairly large feature and currently under review. Once it is merged, we will move to this one.

**Comment by ipapapa** [comment]
#186 has been merged in master, would you mind resolving the conflicts so that we can move forward with this?

**Pull request merged**